File: backend/routes/resale.py
from flask import Blueprint, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
import datetime
from bson.objectid import ObjectId
from dateutil import parser
from routes.health import calculate_vehicle_health

resale_bp = Blueprint("resale", __name__)

def get_db():
    from app import db
    return db

def calculate_resale_report(vehicle, services):
    # 1. Base the initial depreciation on age and original price
    try:
        purchase_price = float(vehicle.get("purchasePrice", 0))
    except (ValueError, TypeError):
        purchase_price = 0

    now = datetime.datetime.now(datetime.timezone.utc)
    
    # Calculate age in years
    age_years = 0
    mfg_year = vehicle.get("manufacturingYear")
    if mfg_year:
        try:
            age_years = now.year - int(mfg_year)
        except:
            pass
            
    # Simple depreciation model: 15% first year, 10% subsequent years
    depreciation_factor = 1.0
    if age_years > 0:
        depreciation_factor *= 0.85
        if age_years > 1:
            depreciation_factor *= (0.90 ** (age_years - 1))
            
    base_value = purchase_price * depreciation_factor
    
    # 2. Trust Score Calculation (Starts at 100)
    trust_score = 100
    trust_factors = []
    
    # Check ownership
    try:
        owners = int(vehicle.get("ownershipCount", 1))
        if owners > 1:
            penalty = (owners - 1) * 5
            trust_score -= penalty
            trust_factors.append({ "type": "negative", "reason": f"Multiple owners ({owners})" })
    except:
        pass
        
    # Process Services for Trust & Value Impact
    verified_services_count = 0
    major_accidents_count = 0
    
    for s in services:
        if s.get("verifiedService"):
            verified_services_count += 1
            
        notes = str(s.get("mechanicNotes", "")).lower()
        cat = str(s.get("serviceCategory", "")).lower()
        if "accident" in notes or "crash" in notes or "major" in cat:
            major_accidents_count += 1
            
    # Apply service trust factors
    if verified_services_count > 0:
        trust_score += min(verified_services_count * 5, 15)
        trust_factors.append({ "type": "positive", "reason": f"Verified service history ({verified_services_count} records)" })
        
    if major_accidents_count > 0:
        trust_score -= (major_accidents_count * 15)
        base_value *= 0.8  # Further 20% depreciation for accidents
        trust_factors.append({ "type": "negative", "reason": f"Major accident/repair history" })
        
    if not services:
        trust_score -= 20
        trust_factors.append({ "type": "negative", "reason": "No service records available" })
        
    # Include Base Health Engine logic
    health_data = calculate_vehicle_health(vehicle, services)
    health_score = health_data.get("healthScore", 100)
    
    # Adjust value based on health
    if health_score >= 85:
        base_value *= 1.05 # 5% premium for excellent health
        trust_factors.append({ "type": "positive", "reason": "Excellent vehicle condition" })
    elif health_score < 70:
        base_value *= 0.9 # 10% penalty for poor health
        
    # Cap trust score
    trust_score = max(0, min(100, trust_score))
    
    # Output Range
    min_value = int(base_value * 0.95)
    max_value = int(base_value * 1.05)
    
    # Define Risk Level
    if trust_score >= 85:
        risk_level = "Low"
    elif trust_score >= 65:
        risk_level = "Medium"
    else:
        risk_level = "High"

    return {
        "vehicleId": str(vehicle.get("_id")),
        "originalPrice": purchase_price,
        "estimatedValueRange": {
            "min": min_value,
            "max": max_value,
            "mean": int(base_value)
        },
        "trustScore": trust_score,
        "trustFactors": trust_factors,
        "maintenanceQuality": health_data.get("conditionLevel", "Good"),
        "riskLevel": risk_level,
        "ageYears": age_years
    }


@resale_bp.route("/<vehicle_id>", methods=["GET"])
@jwt_required()
def get_resale_report(vehicle_id):
    db = get_db()
    vehicles_collection = db["vehicles"]
    services_collection = db["services"]
    owner_id = get_jwt_identity()
    
    try:
        # Verify ownership
        vehicle = vehicles_collection.find_one({
            "_id": ObjectId(vehicle_id),
            "ownerId": owner_id,
            "isArchived": {"$ne": True}
        })
        
        if not vehicle:
            return jsonify({"msg": "Vehicle not found"}), 404
            
        services_cursor = services_collection.find({
            "vehicleId": vehicle_id,
            "isArchived": {"$ne": True}
        })
        services = list(services_cursor)
        
        # Calculate Resale Data
        report_data = calculate_resale_report(vehicle, services)
        
        return jsonify(report_data), 200
        
    except Exception as e:
        return jsonify({"msg": "Error generating resale report", "error": str(e)}), 500

File: backend/src/routes/resale.js
const express = require('express');
const { ObjectId } = require('mongodb');
const { DateTime } = require('luxon');

const { getDb } = require('../db');
const { requireAuth } = require('../middleware/auth');
const { calculateVehicleHealth } = require('./health');

const router = express.Router();

function calculateResaleReport(vehicle, services) {
  const purchasePrice = Number.isNaN(parseFloat(vehicle.purchasePrice)) ? 0 : parseFloat(vehicle.purchasePrice);
  const now = DateTime.utc().startOf('day');

  let ageYears = 0;
  if (vehicle.manufacturingYear) {
    const y = parseInt(vehicle.manufacturingYear, 10);
    if (!Number.isNaN(y)) ageYears = now.year - y;
  }

  let depreciationFactor = 1.0;
  if (ageYears > 0) {
    depreciationFactor *= 0.85;
    if (ageYears > 1) {
      depreciationFactor *= Math.pow(0.9, ageYears - 1);
    }
  }

  let baseValue = purchasePrice * depreciationFactor;

  let trustScore = 100;
  const trustFactors = [];

  // 1. Ownership count penalty
  try {
    const owners = parseInt(vehicle.ownershipCount || 1, 10);
    if (owners > 1) {
      const penalty = (owners - 1) * 5;
      trustScore -= penalty;
      trustFactors.push({ type: 'negative', reason: `Multiple owners (${owners})` });
    }
  } catch (e) {
    // ignore
  }

  // 2. Service history details
  let verifiedServicesCount = 0;
  let majorAccidentsCount = 0;
  for (const s of services || []) {
    if (s.verifiedService) verifiedServicesCount += 1;
    const notes = String(s.mechanicNotes || '').toLowerCase();
    const cat = String(s.serviceCategory || '').toLowerCase();
    if (notes.includes('accident') || notes.includes('crash') || cat.includes('major') || cat === 'accidental repair') {
      majorAccidentsCount += 1;
    }
  }

  // 3. Service records verification
  if (verifiedServicesCount > 0) {
    trustScore += Math.min(verifiedServicesCount * 5, 15);
    trustFactors.push({ type: 'positive', reason: `Verified service history (${verifiedServicesCount} records)` });
  }

  // 4. Major accidents penalty
  if (majorAccidentsCount > 0) {
    trustScore -= majorAccidentsCount * 15;
    baseValue *= 0.8;
    trustFactors.push({ type: 'negative', reason: 'Major accident/repair history' });
  }

  // 5. No service records penalty
  if (!services || services.length === 0) {
    trustScore -= 30;
    trustFactors.push({ type: 'negative', reason: 'No service records available' });
  }

  // 6. Maintenance gaps (if age > 0 and services count is less than ageYears)
  if (services && services.length > 0 && ageYears > 0 && services.length < ageYears) {
    trustScore -= 15;
    trustFactors.push({ type: 'negative', reason: 'Maintenance gaps detected (incomplete service history)' });
  }

  // 7. Vehicle Health / Condition
  const healthData = calculateVehicleHealth(vehicle, services || []);
  const healthScore = healthData.healthScore ?? 100;

  if (healthScore >= 85) {
    baseValue *= 1.05;
    trustScore += 5;
    trustFactors.push({ type: 'positive', reason: 'Excellent vehicle condition' });
  } else if (healthScore >= 70 && healthScore < 85) {
    trustFactors.push({ type: 'positive', reason: 'Good vehicle condition' });
  } else if (healthScore >= 50 && healthScore < 70) {
    baseValue *= 0.9;
    trustScore -= 10;
    trustFactors.push({ type: 'negative', reason: 'Average vehicle condition' });
  } else { // healthScore < 50 (Poor)
    baseValue *= 0.8;
    trustScore -= 25;
    trustFactors.push({ type: 'negative', reason: 'Poor vehicle condition' });
  }

  // 8. Insurance status
  let isInsuranceValid = false;
  if (vehicle && vehicle.insuranceExpiry) {
    const expDate = DateTime.fromISO(vehicle.insuranceExpiry).startOf('day');
    if (expDate.isValid && expDate >= now.startOf('day')) {
      isInsuranceValid = true;
    }
  }

  if (isInsuranceValid) {
    trustFactors.push({ type: 'positive', reason: 'Active insurance policy' });
  } else {
    trustScore -= 5;
    trustFactors.push({ type: 'negative', reason: 'Expired or missing insurance' });
  }

  trustScore = Math.max(0, Math.min(100, trustScore));

  const minValue = Math.trunc(baseValue * 0.95);
  const maxValue = Math.trunc(baseValue * 1.05);

  let riskLevel;
  if (trustScore >= 85) riskLevel = 'Low';
  else if (trustScore >= 65) riskLevel = 'Medium';
  else riskLevel = 'High';

  return {
    vehicleId: String(vehicle._id),
    originalPrice: purchasePrice,
    estimatedValueRange: {
      min: minValue,
      max: maxValue,
      mean: Math.trunc(baseValue)
    },
    trustScore,
    trustFactors,
    maintenanceQuality: healthData.conditionLevel || 'Good',
    riskLevel,
    ageYears
  };
}

router.get('/:vehicle_id', requireAuth, async (req, res) => {
  const vehicleId = req.params.vehicle_id;
  const db = getDb();
  const vehicles = db.collection('vehicles');
  const services = db.collection('services');

  try {
    const vehicle = await vehicles.findOne({
      _id: new ObjectId(vehicleId),
      ownerId: req.user.id,
      isArchived: { $ne: true }
    });

    if (!vehicle) {
      return res.status(404).json({ msg: 'Vehicle not found' });
    }

    const serviceList = await services
      .find({ vehicleId, isArchived: { $ne: true } })
      .toArray();

    const report = calculateResaleReport(vehicle, serviceList);
    return res.status(200).json(report);
  } catch (e) {
    return res.status(500).json({ msg: 'Error generating resale report', error: String(e && e.message ? e.message : e) });
  }
});

module.exports = router;
